src/NuGet.Core/NuGet.Commands/RestoreCommand/LockFileBuilder.cs

```diff
@@ -561,35 +561,45 @@ private IEnumerable<LibraryDependency> GetLibraryDependenciesForCentralTransitiv
         }
 
         /// <summary>
-        /// Enumerates all parent nodes of the specified node.
+        /// Enumerates all inner nodes of the root node which directly or transitively reference the particular graph node.
         /// </summary>
         /// <typeparam name="T">The type of the node.</typeparam>
+        /// <param name="rootNode">The <see cref="GraphNode{TItem}" /> to know which nodes are first level inner nodes.</param>
         /// <param name="graphNode">The <see cref="GraphNode{TItem}" /> to enumerate the parent nodes of.</param>
-        /// <returns>An <see cref="IEnumerable{T}" /> containing a top down list of parent nodes of the specied node.</returns>
-        private static IEnumerable<GraphNode<T>> EnumerateParentNodes<T>(GraphNode<T> graphNode)
+        /// <returns>An <see cref="IEnumerable{T}" /> containing list of parent nodes of the specified node.</returns>
+        private static IEnumerable<GraphNode<T>> EnumerateNodesForDependencyChecks<T>( GraphNode<T> rootNode, GraphNode<T> graphNode)
         {
-            foreach (GraphNode<T> item in graphNode.ParentNodes)
+            var visitedNodes = new HashSet<GraphNode<T>>();
```

Comment:
Yes, indeed. Reusing the `Queue` and the `HashSet` saved noticeable allocations (in my test scenario it was ~200MB.
I tried the same trick for returning results, but this time I didn't notice any significant savings. Thus I think it is better to keep IEnumerable pattern because it is more "elegant" in my opinion.

Reply:
Thanks for profiling that!